src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/TypeCast.cs

```diff
@@ -42,119 +42,147 @@ internal enum AssignmentVariation
-            if (pObjType == pTargetType)
+            Debug.Assert(!pTargetType->IsParameterizedType, "CheckCastClass called with parameterized MethodTable");
+            Debug.Assert(!pTargetType->IsFunctionPointerType, "CheckCastClass called with function pointer MethodTable");
+            Debug.Assert(!pTargetType->IsInterface, "CheckCastClass called with interface MethodTable");
+            Debug.Assert(!pTargetType->HasGenericVariance, "CheckCastClass with variant MethodTable");
+
+            if (obj == null || obj.GetMethodTable() == pTargetType)
             {
                 return obj;
             }
 
-            if (pTargetType->HasGenericVariance && pObjType->HasGenericVariance)
-            {
-                // Only generic interfaces and delegates can have generic variance and we shouldn't see
-                // interfaces for either input here. So if the canonical types are marked as having variance
-                // we know we've hit the delegate case. We've dealt with the identical case just above. And
-                // the regular path below will handle casting to Object, Delegate and MulticastDelegate. Since
-                // we don't support deriving from user delegate classes any further all we have to check here
-                // is that the uninstantiated generic delegate definitions are the same and the type
-                // parameters are compatible.
+            return CheckCastClassSpecial(pTargetType, obj);
+        }
 
-                // NOTE: using general assignable path for the cache because of the cost of the variance checks
-                if (AreTypesAssignableInternal(pObjType, pTargetType, AssignmentVariation.BoxedSource, null))
-                    return obj;
-                return null;
-            }
+        [RuntimeExport("RhTypeCast_CheckCastClassSpecial")]
+        private static unsafe object CheckCastClassSpecial(MethodTable* pTargetType, object obj)
+        {
+            Debug.Assert(!pTargetType->IsParameterizedType, "CheckCastClass called with parameterized MethodTable");
+            Debug.Assert(!pTargetType->IsFunctionPointerType, "CheckCastClass called with function pointer MethodTable");
+            Debug.Assert(!pTargetType->IsInterface, "CheckCastClass called with interface MethodTable");
+            Debug.Assert(!pTargetType->HasGenericVariance, "CheckCastClass with variant MethodTable");
 
-            if (pObjType->IsArray)
+            MethodTable* mt = obj.GetMethodTable();
+            Debug.Assert(mt != pTargetType, "The check for the trivial cases should be inlined by the JIT");
+
+            if (!mt->IsCanonical)
             {
-                // arrays can be cast to System.Object
-                if (WellKnownEETypes.IsSystemObject(pTargetType))
-                {
-                    return obj;
-                }
+                // Arrays should be the only non-canonical types that can exist on GC heap
```

Comment:
> Arrays should be the only non-canonical types that can exist on GC heap

Just curious - is that something that can be changed to not have special cases, or there is a reason for this?

Reply:
We could change the very first `->NonArrayBaseType` to be `->BaseType` and that one will compensate for the arrays. But casting arrays will become slower because obtaining `System.Array` from the correct type universe is slower than the `WellKnownEETypes.IsValidArrayBaseType` check.

Reply:
How does CoreCLR store both base type and the element of the array in the `MethodTable`?

I don't see a way to put both base and element type into the `MethodTable` without either leaving empty space in non-array `MethodTable`s, or placing the element type in the variable-sized portion of it (after the virtual slots). Placing it after virtual slots makes accessing array element type slower so that would be another compromise and that feels like too many compromises.

Reply:
CoreCLR stores element type in a union that is also used for things that are not applicable to arrays. The base is stored the same as with other types though, so no specialcasing for arrays.

```c
    union
    {
        PerInstInfo_t m_pPerInstInfo;
        TADDR         m_ElementTypeHnd;
        TADDR         m_pMultipurposeSlot1;
    };
```

Reply:
I do not see an obvious place to store the array element type in a NativeAOT method table.

Reply:
Yeah, CoreCLR's `MethodTable` has a bunch of extra stuff in it because the type system serves both execution and compilation/loading. NativeAOT is execution-only.

I think if we're concerned about the extra check, we could have casts to `System.Object` and `System.Array` take the slower path and the check won't be necessary. Casting to `Object` is hypothetical. Casting to `Array` sometimes happens. And if slower path not feasible, we could introduce yet another flavor of the fast helper.

I don't think it's so concerning at this point.